javascript/ql/src/semmle/javascript/security/dataflow/ZipSlip.qll

```diff
@@ -0,0 +1,124 @@
+/**
+ * Provides a taint tracking configuration for reasoning about unsafe zip extraction.
+ */
+
+import javascript
+
+module ZipSlip {
+  /**
+   * A data flow source for unsafe zip extraction.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for unsafe zip extraction.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer guard for unsafe zip extraction.
+   */
+  abstract class SanitizerGuard extends TaintTracking::SanitizerGuardNode, DataFlow::ValueNode { }
+
+  /** A taint tracking configuration for Zip Slip */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "ZipSlip" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isSanitizerGuard(TaintTracking::SanitizerGuardNode nd) {
+      nd instanceof SanitizerGuard
+    }
+  }
+
+  /**
+   * Holds if `node1` flows to `node2` in one step by virtue of
+   * `node2` being of the form `.pipe(node1)`. The reason this flow
+   * exists is that `.pipe` returns its argument to make chained
+   * stream operations work.
+   */
+  predicate pipeStep(DataFlow::Node node1, DataFlow::MethodCallNode node2) {
```

Comment:
We definitely want these streaming/piping flowsteps in the standard library. Can we make them all private in this file for now? Then we can move them without breaking backwards compatibility later.

Reply:
Is there a performance-acceptable way of putting this in the standard library? I'd absolutely be interested in doing that in a separate PR, but I don't see how, yet, per @asger-semmle's comments. In any event I took care to make the auxiliary predicates I have private.

Reply:
Let us look at that after the merge.